Add initial version of CoMEt \Phi in Python

mdml opened the pull request:
The goal of this pull request is to have a Python implementation of the CoMEt \Phi that calls the corresponding C functions to compute the P-value. The main purpose of \Phi is to decide when to use an  approximation vs. the exact test, but currently both the permutational approximation and \Phi implementations in C require the user to pass in ALL the mutation data, which I think is too much of a burden.

Here I've implemented a pure Python version of the permutational, and a Python version of \Phi that calls the C versions of the exact and binomial tests, and the pure Python permutational.

The permutational looks good to me, but there are several issues with \Phi.
1. The following example doesn't work (returns a P-value of 0), which could be a precision issue or something more nefarious.
   
   ```
   >>> from comet import phi
   >>> phi(4, 289, [176, 30, 27, 0, 30, 0, 0, 0, 22, 0, 1, 0, 0, 1, 0, 2])
   (0.0, 'e')
   ```
2. The permutational may return 0, which seems potentially confusing and definitely undesirable. I'm not sure how to make this match the C implementation.
3. I currently have the \Phi return a tuple with P-value and a character representing which test (exact, binomial, or permutational) was used to compute the P-value. Is this what we want \Phi to return?

Commit:
Subtype Parser

Initial script for adding streamlining subtype information as well as
core events, with test input.

Commit:
Adding sample inputs

Sample inputs.

Commit:
Fixed subtype inclusion criteria

Commit:
Adding in subtype information for samples without subtype info

Commit:
Subtype parser integration

Commit:
add initial version of CoMEt phi in Python

Commit:
Subtype parser fully integrated

Fully integrated and verified on STAD gastric data:
http://compbio-research.cs.brown.edu/comet-test/

Commit:
Mutation data and subtype processing update

- Changed a line with feedback from @mdml
- Removed an old comment line that doesn’t seem to be doing
anything/necessary
- Added comments in run_comet.py

bournewu committed:
Merge pull request #14 from raphael-group/subtype-parser

Subtype Parser

bournewu commented on comet/__init__.py:
```diff
@@ -1,5 +1,58 @@
+        if exact_pval != -1: # the P-value was computed exactly
+            return exact_pval, EXACT
+        else: # the P-value never finished
+            return py_permutation_test(k, N, tbl, num_permutations), PERMUTATIONAL
```
Here return the binomial if the permutation P-value is 0.

braphael8 replied:
Is this the right behavior?  Shouldn't the permutation p-value return
1/#perm instead of zero?  Or possibly the min of 1/#perm and the binomial
approximation?

On Tuesday, February 9, 2016, Hsin-Ta Wu notifications@github.com wrote:

> In comet/**init**.py
> https://github.com/raphael-group/comet/pull/15#discussion_r52340765:
> 
> > -        # minimum number of co-occurrences (since the approximation is poor
> > -        # for mostly exclusive cases) and (b) the P-value is above a minimum
> > -        # value
> > -        binom_pval = binom_test(k, N, list(tbl), 1.1)
> > -        conum = sum(tbl[1:]) - sum(tbl[2**i] for i in range(k))
> > -        if conum > co_cutoff or binom_pval > binom_pvalthresh:
> > -            return binom_pval, BINOM
> >   +
> > -        # If the binomial doesn't work, then try the exact test. The exact test
> > -        # runs until either the P-value is computed, or accumulated mass
> > -        # reaches a certain value.
> > -        num_tbls, exact_pval = exact_test(k, N, tbl, 1.1)
> > -        if exact_pval != -1: # the P-value was computed exactly
> > -            return exact_pval, EXACT
> > -        else: # the P-value never finished
> > -            return py_permutation_test(k, N, tbl, num_permutations), PERMUTATIONAL
> 
> Here return the binomial if the permutation P-value is 0.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/raphael-group/comet/pull/15/files#r52340765.

mdml replied:
I don't think it should return 1/#perm, since it really should be P < 1/#perm. I think we can leave it up to the user to figure out how to handle those types of P-values? In our case, we often prefix the P-values with "<" so it's clear.

@braphael8: @bournewu: What do you think?

bournewu replied:
The P-values of permutation test will only be used if there is at least one random permutation has more exclusive mutations than the observed one, otherwise it will perform binomial approximation. Thus, the P-Values with the prefix "< z+1/#perm" here looks good for me, where z is the number of permutations with more exclusive mutations than the observed one.

Commit:
Merge branch 'py_comet_phi' of https://github.com/raphael-group/comet into py_comet_phi

Commit:
use the binomial when the permutational returns 0

Commit:
use the exact_pvalthresh in phi

mdml commented:
I updated `phi` to return the binomial when the permutation P-value is 0. This is currently what the C code does, so I think this is what the Python code should do for now unless we decide to change both.

I also updated the exact test to use the `exact_pvalthresh`, which was an important bug.

Finally, @bournewu the example from (1) above is now working for me:

```
In [1]: from comet import *
In [2]: precompute_factorials(289)
In [3]: phi(4, 289, [176, 30, 27, 0, 30, 0, 0, 0, 22, 0, 1, 0, 0, 1, 0, 2])
Out[3]: (9.828128648363476e-05, 'e')
```

Can you verify? I think the earlier error may have something to do with not precomputing the factorials and not using the `exact_pvalthresh`.

bournewu commented:
Codes look good to me. And I also tested a gene set CDKN2A(D), CDK4(A), MSL3, RB1 in GBM and it worked great.

``` python
phi(4, 261, [28,4,16,0,37,0,0,0,156,1,3,0,16,0,0,0], 0.001, 0.005, 20, 1000)
(2.1632398806774606e-21, 'e')
```
